test/datoms_differ/api_test.cljc

```diff
@@ -0,0 +1,492 @@
+                  [:db/add 1026 :service/id :s567]
+                  [:db/add 1027 :route/number "100"]
+                  [:db/add 1027 :route/services 1026]
+                  [:db/add 1027 :route/services 1028]
+                  [:db/add 1028 :service/id :s789]
+                  [:db/add 1028 :service/trips 1025]
+                  [:db/add 1028 :service/trips 1029]
+                  [:db/add 1029 :trip/id "bar"]]
+        :eavs #{(d/datom 1024 :vessel/imo "123" :prep)
+                (d/datom 1025 :trip/id "foo" :prep)
+                (d/datom 1029 :trip/id "bar" :prep)
+                (d/datom 1026 :service/allocated-vessel 1024 :prep)
+                (d/datom 1026 :service/id :s567 :prep)
+                (d/datom 1027 :route/number "100" :prep)
+                (d/datom 1027 :route/services 1026 :prep)
+                (d/datom 1027 :route/services 1028 :prep)
+                (d/datom 1028 :service/id :s789 :prep)
+                (d/datom 1028 :service/trips 1025 :prep)
+                (d/datom 1028 :service/trips 1029 :prep)}
+        :refs {[:route/number "100"] 1027
+               [:service/id :s567] 1026
+               [:service/id :s789] 1028
+               [:trip/id "foo"] 1025
+               [:trip/id "bar"] 1029
+               [:vessel/imo "123"] 1024}}))
+
+    (testing "empty db - nil val for attr throws"
+      (is (thrown? Exception
+                   (sut/with (db) :prepare-routes [{:route/number "100"
+                                                    :route/name nil}]))))
+
+    (testing "Running out of eids for db-id-partition throws"
+      (let [db (sut/create-conn (assoc schema ::sut/db-id-partition {:from 1 :to 2}))]
+        (try
+          (sut/with @db :prep [{:vessel/imo "123"}
+                               {:vessel/imo "234"}
+                               {:vessel/imo "345"}])
+          (is (= :should-throw :didnt))
+          (catch Exception e
+            (is (str/starts-with? (.getMessage e) "Generated internal eid falls outside internal db-id-partition,"))))))
+
+    (testing "empty db - entity without identity attributes throws"
+      (is (thrown? Exception
+                   (sut/with (db) :prepare-routes [{}])))
+      (is (thrown? Exception
+                   (sut/with (db) :prepare-routes [{:route/name "Dufus"}]))))
+
+    (testing "conflicting values for entity attr throws"
+      (try
+        (sut/with (db) :prepare-routes [{:route/number "100" :route/name "Stavanger-Taua"}
+                                        {:route/number "100" :route/name "Stavanger-Tau"}])
+        (is (= :should-throw :didnt))
+        (catch Exception e
+          (is (= "Conflicting values asserted for entity" (.getMessage e)))
+          (is (= {:entity-ref [:route/number "100"]
+                  :attr :route/name
+                  :conflict {:prepare-routes #{"Stavanger-Taua" "Stavanger-Tau"}}}
+                 (ex-data e)))))))
```

Comment:
Checking the ex-data is great. I don't think there's a corresponding `thrown-with-ex-data?`, so we should probably keep this construct here. :)